### PR TITLE
[Kunlun] add gen_bkcl_id_op, support multi XPU cards training using multiprocess

### PR DESCRIPTION
**Pull request by vslyu**
### PR types

New features
### PR changes

Others
### Describe

-  add gen_bkcl_id_op for multi Baidu Kunlun cards training
-  support fleet api for multi Baidu Kunlun cards training

### CODE REVIEW

File: paddle/fluid/operators/collective/gen_bkcl_id_op.cc [hunk 1]
```diff
@@ -0,0 +1,194 @@
+/* Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wangxicoding) [inline paddle/fluid/operators/collective/gen_bkcl_id_op.cc:1]: 2021

File: python/paddle/fluid/tests/unittests/test_dist_base.py [hunk 2]
```diff
@@ -566,12 +574,12 @@ def run_gpu_fleet_api_trainer(self, args):
         args.trainer_id = paddle.distributed.get_rank()
 
         # 3. init parallel env
-        if args.update_method == "nccl2":
+        if args.update_method == "nccl2" or "bkcl":
```
Comment (wangxicoding) [inline python/paddle/fluid/tests/unittests/test_dist_base.py:577]: this fleet is paddle.fluid.incubate.fleet, maybe need another test

File: python/paddle/fluid/tests/unittests/xpu/test_gen_bkcl_id_op.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+import os
+import copy
+import sys
+sys.path.append("..")
```
Comment (wangxicoding) [inline python/paddle/fluid/tests/unittests/xpu/test_gen_bkcl_id_op.py:19]: from ..launch_function_helper

**Review by wangxicoding — COMMENTED** [review]
后续TODO建议 @vslyu 
1、统一gen_nccl_id、c_gen_nccl_id、gen_bkcl_id、c_gen_bkcl_id，叫c_gen_comm_id，采用kernel op的形式。
2、c_comm_init_op，采用kernel op的形式。

File: python/paddle/fluid/tests/unittests/test_dist_base.py (hunk 2 above)
Comment (vslyu) [reply python/paddle/fluid/tests/unittests/test_dist_base.py:577]: New paddle.distributed.fleet API interface is  imported again In this run_use_fleet_api_trainer function:
```
import paddle.distributed.fleet as fleet
import paddle.distributed.fleet.base.role_maker as role_maker
```
so , the unittest finally runs with new paddle.distributed.fleet API.

File: paddle/fluid/operators/collective/c_comm_init_op.cc [hunk 4]
```diff
@@ -39,29 +42,56 @@ class CCommInitOp : public framework::OperatorBase {
...
                const platform::Place& place) const override {
-    PADDLE_ENFORCE_EQ(is_gpu_place(place), true,
+    PADDLE_ENFORCE_EQ(is_gpu_place(place) || is_xpu_place(place), true,
                       platform::errors::PreconditionNotMet(
-                          "CCommInitOp can run on gpu place only."));
+                          "CCommInitOp can run on gpu or xpu place only."));
 
     auto var = scope.FindVar(Input("X"));
     PADDLE_ENFORCE_NOT_NULL(
         var, platform::errors::InvalidArgument("Input con not be empty."));
```
Comment (chenwhql) [inline paddle/fluid/operators/collective/c_comm_init_op.cc:51]: `con` -> `can`?

File: paddle/fluid/operators/collective/c_comm_init_op.cc [hunk 5]
```diff
@@ -39,29 +42,56 @@ class CCommInitOp : public framework::OperatorBase {
...
+      if (Attr<int>("device_id") >= 0) {
+        device_id = Attr<int>("device_id");
+      }
+      platform::NCCLCommContext::Instance().CreateNCCLComm(
+          nccl_id, nranks, rank_id, device_id, rid);
 #else
-    PADDLE_THROW(platform::errors::PreconditionNotMet(
-        "PaddlePaddle should compile with GPU."));
+      PADDLE_THROW(platform::errors::PreconditionNotMet(
+          "PaddlePaddle should compile with GPU."));
```
Comment (chenwhql) [inline paddle/fluid/operators/collective/c_comm_init_op.cc:67]: `should compile` -> `should be compiled`?

File: paddle/fluid/operators/collective/c_comm_init_op.cc [hunk 6]
```diff
@@ -39,29 +42,56 @@ class CCommInitOp : public framework::OperatorBase {
...
+              ring_id));
+      int device_id = BOOST_GET_CONST(platform::XPUPlace, place).device;
+      if (Attr<int>("device_id") >= 0) {
+        device_id = Attr<int>("device_id");
+      }
+      platform::BKCLCommContext::Instance().CreateBKCLComm(
+          bkcl_id, nranks, rank_id, device_id, rid);
+#else
+      PADDLE_THROW(platform::errors::PreconditionNotMet(
+          "PaddlePaddle should compile with XPU."));
```
Comment (chenwhql) [inline paddle/fluid/operators/collective/c_comm_init_op.cc:89]: same above

**Review by chenwhql — APPROVED** [review]
LGTM for PADDLE_ENFORCE, comment can be fixed in next PR

File: python/paddle/fluid/tests/unittests/xpu/test_gen_bkcl_id_op.py (hunk 3 above)
Comment (vslyu) [reply python/paddle/fluid/tests/unittests/xpu/test_gen_bkcl_id_op.py:19]: 保留sys.path.append("..")，保持和其他xpu单测一致吧

File: paddle/fluid/operators/collective/gen_bkcl_id_op.cc (hunk 1 above)
Comment (vslyu) [reply paddle/fluid/operators/collective/gen_bkcl_id_op.cc:1]: 下个pr改掉.

**Comment by vslyu** [comment]
> 后续TODO建议 @vslyu
> 1、统一gen_nccl_id、c_gen_nccl_id、gen_bkcl_id、c_gen_bkcl_id，叫c_gen_comm_id，采用kernel op的形式。
> 2、c_comm_init_op，采用kernel op的形式。

收到，待xpu动态图，静态图性能优化时再考虑精简代码。

**Review by wangxicoding — APPROVED** [review]

File: paddle/fluid/operators/collective/c_comm_init_op.cc (hunk 6 above)
Comment (vslyu) [reply paddle/fluid/operators/collective/c_comm_init_op.cc:89]: 收到，下个pr改掉.

File: paddle/fluid/operators/collective/c_comm_init_op.cc (hunk 5 above)
Comment (vslyu) [reply paddle/fluid/operators/collective/c_comm_init_op.cc:67]: 收到，下个pr改掉.

**Review by luotao1 — APPROVED** [review]
LGTM for framework.py

File: paddle/fluid/operators/collective/c_comm_init_op.cc (hunk 4 above)
Comment (vslyu) [reply paddle/fluid/operators/collective/c_comm_init_op.cc:51]: 收到，下个pr改掉.

**Review by phlrain — APPROVED** [review]

**Pull request merged**